{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Editor : AJ\n",
    "\n",
    "######### CyREST - R implementation #########\n",
    "\n",
    "# Document and resource used : \n",
    "#\n",
    "#        -RCy3: Network biology using Cytoscape from within R ref.\n",
    "#         https://f1000research.com/articles/8-1774/v1\n",
    "#        -RCy3 Cytoscape functions ref.\n",
    "#         https://cytoscape.org/RCy3/reference/index.html\n",
    "#        -cytoscape-automation ref.\n",
    "#         https://github.com/cytoscape/cytoscape-automation/blob/master/for-scripters/R/notebooks/Cancer-networks-and-data.Rmd\n",
    "#        -RCy3: Functions to Access and Control Cytoscape ref.\n",
    "#         https://rdrr.io/bioc/RCy3/\n",
    "#        -https://cytoscape.org/cytoscape-tutorials/presentations/bioc2018_Rcy3_intro.html#/\n",
    "#        -bioconductor RCy3 Documentation ref.\n",
    "#         https://bioconductor.org/packages/release/bioc/html/RCy3.html\n",
    "#        -clusterMaker2: Creating and Visualizing Cytoscape Clusters ref.\n",
    "#         https://www.rbvi.ucsf.edu/cytoscape/clusterMaker2/clusterMaker2.shtml#mcl\n",
    "#        -AutoAnnotate Cytoscape App 1.3 ref.\n",
    "#         https://autoannotate.readthedocs.io/en/latest/index.html\n",
    "#\n",
    "\n",
    "############\n",
    "# purpose : Creating analysis network from proteogenomic and protein-protein interactions' information \n",
    "#           obtained from differential expression, gene ontology, pathway analysis and STRING database\n",
    "#           queries\n",
    "# parameters : csv files (containing proteogenomic and proteomics interactions' information ),\n",
    "#              number of conditions, species, ontology domain, network creation attributes,\n",
    "#              visual styles attributes, clustring attributes, auto-annotation attributes,\n",
    "#              network's layput attributes\n",
    "# return value : \n",
    "#                - Summary of network and clustering evaluation\n",
    "#                -\n",
    "#                \n",
    "# details :\n",
    "#        - Cytoscape must be running whenever using RCy3 is used\n",
    "#        - Initial setting such as Species, ontology domain, addrress of the directory which contains proteogenomic information,\n",
    "#        visual styles attributes, mcl clustring attributes, and annotation attributes should be provided in the section : \n",
    "#        ##  Setting initil Analysis' info  ##\n",
    "#        - Helpful Cytoscape commands :\n",
    "#                commandsAPI(), commandsHelp(\"app name\"), getLayoutNames(),\n",
    "#                commandsRun('app name'), commandsHelp(\"app name subcommand\")\n",
    "#\n",
    "############\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "############################################\n",
    "#            Network analysis :\n",
    "############################################ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading needed Libraries\n",
    "\n",
    "libs = c(\"RCy3\", \"RColorBrewer\")\n",
    "\n",
    "for (i in libs){\n",
    "  if( !is.element(i, .packages(all.available = TRUE)) ) {\n",
    "    install.packages(i)\n",
    "  }\n",
    "  library(i,character.only = TRUE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install needed Cytoscape apps\n",
    "\n",
    "# App list\n",
    "cyto_app_l <- c(\"clustermaker2\", \"autoannotate\", \"wordcloud\", \"Legend Creator\")\n",
    "\n",
    "# check if the apps are already installed\n",
    "for(idx in 1:length(cyto_app_l)){\n",
    "    \n",
    "    check <- grep(commandsGET(paste(\"apps status app=\\\"\", cyto_app_l[idx],\"\\\"\", sep=\"\")),\n",
    "                    pattern = \"status: Installed\")\n",
    "    \n",
    "    # install app if it is not installed\n",
    "    if(length(check) == 0){\n",
    "        print(cyto_app_l[idx])\n",
    "        installApp(cyto_app_l[idx])\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating related directories\n",
    "\n",
    "dir.create(file.path(\"../../Results/\",\"Cy_networks\"), recursive = TRUE, showWarnings = FALSE)\n",
    "dir.create(file.path(\"../../Results/\",\"Cy_manual_networks\"), recursive = TRUE, showWarnings = FALSE)\n",
    "dir.create(file.path(\"../../Results/\",\"Cy_network_images\"), recursive = TRUE, showWarnings = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creatnig a 'named list' containing mapping between NCBI species Ids and\n",
    "# supported species to be used for downloading proteomics data from STRING \n",
    "\n",
    "NCBI_id_list <- list(9606,10088,10114,4932,7227,3701,7955,10224,27592,9031,9615,9823,\n",
    "                     9544,83333,8353,7164,9598,36329,83334,32)\n",
    "\n",
    "list_names <- c(\"human\",\"mouse\",\"rat\",\"yeast\",\"fly\",\"arabidopsis\",\"zebrafish\",\"worm\",\"bovinae\",\n",
    "                \"chicken\",\"canine\",\"pig\",\"rhesus\",\"ecoli_k12\",\"xenopus\",\"anopheles\",\"chimp\",\n",
    "                \"malaria\",\"ecoli_sakai\",\"myxococcus\")\n",
    "\n",
    "names(NCBI_id_list) <- list_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##         Setting initil info          ##\n",
    "##########################################\n",
    "\n",
    "# define the prfered species' name from the names in list_names \n",
    "species = \"fly\"\n",
    "\n",
    "# define the prfered ontology domain\n",
    "ont_domain = \"biological_process\"           #  possible values : \"cellular_component\",\"molecular_function\",\n",
    "                                                                                                                        #\" biological_process\"\n",
    "##     Setting the input data info      ##\n",
    "##########################################\n",
    "\n",
    "# Defining number of conditions (treatments) except the base condition,\n",
    "treat_count = 3\n",
    "\n",
    "##     Network creation attributes      ##\n",
    "##########################################\n",
    "id_name = \"preferredName\"               # Name of nodes' input table column being used as\n",
    "                                        # nodes' input table id\n",
    "interaction_source = \"preferredName_A\"  # Name of edges(interaction)' input table column being used as\n",
    "                                        # edges' table source (Example of interaction input table can be \n",
    "                                        # the result of an query from STRING database)\n",
    "interaction_target = \"preferredName_B\"  # Name of edges(interaction)' input table column being used as\n",
    "                                        # edges' table target (Example of interaction input table can be \n",
    "                                        # the result of an query from STRING database)\n",
    "interaction_weight = \"combined_score\"   # Name of edges(interaction)' input table column being used as\n",
    "                                        # edges' table weight (Example of interaction input table can be \n",
    "                                        # the result of an query from STRING database)\n",
    "\n",
    "## Setting the visual styles attributes ##\n",
    "##########################################\n",
    "BackgroundColor = '#888888'          # Hexadecimal8-bit per channel \n",
    "NodeShape = \"ellipse\"                # Use getArrowShapes() to see the list of shapes\n",
    "NodeSize = 60                        # Numeric\n",
    "NodeDefaultColor = \"#AAAAAA\"         # Hexadecimal8-bit per channel\n",
    "EdgeDefaultLineWidth = 2\n",
    "\n",
    "size_col = \"F\"                       # Numeric column - Name of nodes' table column being used \n",
    "                                     # to map the size of the nodes\n",
    "color_col = \"score\"                  # Numeric column - Name of nodes' table column being used\n",
    "                                     # to map the color of the nodes\n",
    "NodeLabel = \"preferredName\"          # Name of nodes' table column being used to get the node names from\n",
    "\n",
    "##   setting mcl clustring attributes   ##\n",
    "##########################################\n",
    "edgeCutOff = 0.4                     # Allows the user to set a custoff for edge weights\n",
    "                                     # in MC clustering very dense networks\n",
    "edgeWeighter = \"None\"                # Conversions that might be applied to the edge weights before clustering.\n",
    "                                     # list values : \n",
    "                                     #      None: Don't do any conversion\n",
    "                                     #      1/value: Use the inverse of the value.\n",
    "                                     #      LOG(value): Take the log of the value.\n",
    "                                     #      -LOG(value): Take the negative log of the value.\n",
    "                                     #      SCPS: (Spectral Clustering of Protein Sequences) algorithm\n",
    "inflation_parameter = 2.5            # The Granularity Parameter is also known as the inflation \n",
    "                                     # parameter or inflation value.This is the power used to \n",
    "                                     # inflate the matrix. Reasonable values range from ~1.8 to about 2.5\n",
    "\n",
    "##   setting auto annotate attributes   ##\n",
    "##########################################\n",
    "\n",
    "clusterIdColumn = \"__mclCluster\"     # Allows to define clusters using node attributes for AutoAnnotate \n",
    "                                     # This allows clustering algorithms provided by other Cytoscape Apps \n",
    "                                     # or by external scripts to be used with AutoAnnotate.\n",
    "createSingletonClusters = \"false\"    # If \"true\" a cluster will be created for each un-clustered node.\n",
    "maxWords = 3                         # Number of words to be extrcted by annotation algorithm and used for annotation\n",
    "useClusterMaker = \"false\"            # Alows AutoAnnotate to use other cytoscape apps for clustering the nodes\n",
    "\n",
    "labelColumn = \"pathways\"              # Select a Node Column that will be used to calculate the cluster labels.\n",
    "                                                                                       # possible values : \"GO_Term\", \"pathways\"\n",
    "\n",
    "## setting network's layput attributes  ##\n",
    "##########################################\n",
    "layout = \"force-directed\" # other possible lay outs : attribute-circle, stacked-node-layout,\n",
    "                          # degree-circle, circular , attributes-layout, kamada-kawai,\n",
    "                          # force-directed, cose, grid, hierarchical, fruchterman-rheingold,\n",
    "                          # isom\n",
    "\n",
    "defaultSpringCoefficient = 0.00008\n",
    "defaultSpringLength = 100\n",
    "\n",
    "##  setting the results' output format  ##\n",
    "##########################################\n",
    "\n",
    "image_type = \"SVG\"           # PNG (default), JPEG, PDF, SVG, PS (PostScript).\n",
    "network_type = \"SIF\"         # File type. SIF (default), CX, cyjs, graphML, NNF, xGMML. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing Cytoscape connection\n",
    "\n",
    "base.url = \"http://localhost:1234\"\n",
    "cytoscapeVersionInfo ()\n",
    "cytoscapePing ()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load data from file\n",
    "\n",
    "ending = paste(species, ont_domain, sep = \"_\")\n",
    "\n",
    "# Load Proteogenomic information tabels \n",
    "\n",
    "Proteogenomic_dfl = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    Proteogenomic_dfl[[idx]] <- read.table(file = paste(\"../../Results/Proteogenomic_info/\",\n",
    "                                                      ending, \"_proteogenomic_info_\", idx, \".csv\", sep = \"\"),\n",
    "                                  ,sep = \",\", header = TRUE,  stringsAsFactors = FALSE)\n",
    "    Proteogenomic_dfl[[idx]]$score <- Proteogenomic_dfl[[idx]]$F * Proteogenomic_dfl[[idx]]$regulation\n",
    "}\n",
    "\n",
    "# Load Proteomics interaction tables' data\n",
    "\n",
    "interaction_df = list()\n",
    "node_id = list()\n",
    "target_id = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    interaction_df[[idx]] <- read.table(file = paste(\"../../Results/Proteomics_interaction_tables/\",\n",
    "                                                 ending, \"_Proteomics_interaction_table_\", idx, \".csv\", sep = \"\"),\n",
    "                                  ,sep = \",\", header = TRUE,  stringsAsFactors = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating intial proteomics interaction networks using imported tables' data\n",
    "\n",
    "network_id = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    nodes <- data.frame(id=Proteogenomic_dfl[[idx]][,paste(id_name)],\n",
    "                        Proteogenomic_dfl[[idx]],\n",
    "                        stringsAsFactors=FALSE)\n",
    "    edges <- data.frame(source=interaction_df[[idx]][,paste(interaction_source)],\n",
    "                        target=interaction_df[[idx]][,paste(interaction_target)],\n",
    "                        interaction_df[[idx]],\n",
    "                        #interaction = interaction_list # optional to provide a list interaction types\n",
    "                        weight=interaction_df[[idx]][,paste(interaction_weight)], # numeric\n",
    "                        stringsAsFactors=FALSE)\n",
    "    \n",
    "    network_id[[idx]] <- createNetworkFromDataFrames(nodes, edges, \n",
    "                                                      title = paste(\"Proteogenomic_net_\", idx, sep = \"\"),\n",
    "                                                      collection = \"String_DataFrames\")\n",
    "    #layoutNetwork('force-directed')\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Definning the visualiziation expression and setting the network Style\n",
    "\n",
    "# Define new visual styles and style's default values\n",
    "\n",
    "style.name = \"new_style\"\n",
    "createVisualStyle(style.name)\n",
    "setBackgroundColorDefault(BackgroundColor, style.name)\n",
    "setNodeShapeDefault(NodeShape, style.name) \n",
    "setNodeSizeDefault(NodeSize, style.name)\n",
    "setNodeColorDefault(NodeDefaultColor, style.name)\n",
    "setEdgeLineWidthDefault(EdgeDefaultLineWidth, style.name)\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    \n",
    "    # Choosing the network\n",
    "    setCurrentNetwork(network = getNetworkName(suid=as.numeric(network_id[[idx]])))\n",
    "    \n",
    "    #Setting new visual styles\n",
    "    setVisualStyle(style.name)\n",
    "    \n",
    "    # Get the min and max differentiation score from node table and defining the range of values\n",
    "    # of visualiziation size node expression data mapping.\n",
    "    \n",
    "    size = getTableColumns(\"node\", size_col)\n",
    "    size.values = c(min(size[,1],na.rm=TRUE),mean(size[,1],na.rm=TRUE),max(size[,1],na.rm=TRUE))\n",
    "    \n",
    "    # Get the min and max protein interaction score from edge table and defining the range of values\n",
    "    # of visualiziation the edge line width expression data mapping .\n",
    "    \n",
    "    edge_weight = getTableColumns('edge', \"combined_score\" )\n",
    "    min.edge_weight = min(edge_weight[,1],na.rm=TRUE)\n",
    "    max.edge_weight = max(edge_weight[,1],na.rm=TRUE)\n",
    "    edge_weight.values = c(min.edge_weight,\n",
    "                           (min.edge_weight+max.edge_weight)*0.5,\n",
    "                           max.edge_weight)\n",
    "\n",
    "    # Get the min and max differentiation score from node table and defining the range of values\n",
    "    # of visualiziation color node expression data mapping .\n",
    "    \n",
    "    node_col = getTableColumns('node', color_col )\n",
    "    \n",
    "    min.node_col = min(node_col[,1],na.rm=TRUE)\n",
    "    max.node_col = max(node_col[,1],na.rm=TRUE)\n",
    "    color.data.values = c(min.node_col,0,max.node_col)\n",
    "    \n",
    "    # picking colors with RColorBrewer package\n",
    "    # display.brewer.all(length(data.values), colorblindFriendly=TRUE, type=\"div\") # div,qual,seq,all\n",
    "    node.colors <- c(rev(brewer.pal(length(color.data.values), \"RdBu\")))\n",
    "    \n",
    "    # constructing CyREST data objects for style mappings \n",
    "    # Mapping node size\n",
    "    setNodeSizeMapping(size_col , size.values, \n",
    "                       sizes = c(30, 70, 150), mapping.type = \"c\", \n",
    "                       style.name = style.name)\n",
    "    # Mapping node labels\n",
    "    setNodeLabelMapping(NodeLabel, style.name)\n",
    "    # Mapping node colors\n",
    "    setNodeColorMapping(color_col, color.data.values, node.colors,\n",
    "                        style.name=style.name)\n",
    "    # Mapping edge line width\n",
    "    setEdgeLineWidthMapping(\"combined_score\", edge_weight.values, widths = c(1,3,6),\n",
    "                            mapping.type = \"c\", style.name=style.name)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Clustring with mcl Algorithm\n",
    "\n",
    "new_node_table = list()\n",
    "new_edge_table = list()\n",
    "\n",
    "# setting mcl clustring attributes\n",
    "\n",
    "adjustLoops = \"true\"\n",
    "undirectedEdges = \"true\"\n",
    "selectedOnly = \"false\"\n",
    "createGroups = \"true\"\n",
    "showUI = \"true\"\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "        \n",
    "    # Choosing the network\n",
    "    network = paste(\"Proteogenomic_net_\", idx, sep = \"\")\n",
    "    setCurrentNetwork(network = getNetworkName(suid=as.numeric(network_id[[idx]])))\n",
    "    \n",
    "    # creating clustering command\n",
    "    clustermaker_url <- paste(\"cluster mcl edgeCutOff=\", edgeCutOff, \"network=\", network, \n",
    "                              \"edgeWeighter=\", edgeWeighter, \"adjustLoops=\", adjustLoops, \n",
    "                              \"inflation_parameter=\", inflation_parameter, \"edgeWeighter=\",\n",
    "                              edgeWeighter, \"showUI=\", showUI, \"undirectedEdges=\",\n",
    "                              undirectedEdges)\n",
    "    \n",
    "    # running clustring command\n",
    "    # commandsRun(clustermaker_url)\n",
    "    commandsGET(clustermaker_url)\n",
    "    \n",
    "    #get the clustering results\n",
    "    network = paste(\"Proteogenomic_net_\", idx, \"--clustered\", sep = \"\")\n",
    "    new_node_table[[idx]] <- getTableColumns(table= \"node\",network = as.numeric(network_id[[idx]]))\n",
    "    new_edge_table[[idx]] <- getTableColumns(table= \"edge\",network = as.numeric(network_id[[idx]]))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Applying network layput\n",
    "\n",
    "if (TRUE){\n",
    "    \n",
    "    for (idx in 1:treat_count){\n",
    "        \n",
    "        # Choosing the network\n",
    "        network = paste(\"Proteogenomic_net_\", idx, \"--clustered\", sep = \"\")\n",
    "        setCurrentNetwork(network = network)\n",
    "        # Creating layout command\n",
    "        \n",
    "        # constructing CyREST data object for style mappings \n",
    "        if(layout == \"force-directed\"){\n",
    "            \n",
    "            layout <- paste(layout, \"defaultSpringCoefficient=\",\n",
    "                              defaultSpringCoefficient,\n",
    "                              \"defaultSpringLength=\",\n",
    "                              defaultSpringLength)\n",
    "            layoutNetwork(layout)\n",
    "            }\n",
    "        else {\n",
    "            layoutNetwork(layout)\n",
    "        }\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Preforming auto-annotation\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    \n",
    "    # Choosing the network\n",
    "    network = paste(\"Proteogenomic_net_\", idx, \"--clustered\", sep = \"\")\n",
    "    setCurrentNetwork(network = network)\n",
    "    \n",
    "    # Add list of words to be ignored during annotation \n",
    "    \n",
    "    wordcloud_ignore1 <- paste(\"wordcloud ignore add\", \"value=\", \"null\")\n",
    "    wordcloud_ignore2 <- paste(\"wordcloud ignore add\", \"value=\", \"na\")\n",
    "\n",
    "    commandsGET(wordcloud_ignore1)\n",
    "    commandsGET(wordcloud_ignore2)\n",
    "\n",
    "    # Run the AutoAnnotate command\n",
    "    annotate <- paste(\"autoannotate annotate-clusterBoosted\", \"network=\", network,\n",
    "                      \"clusterIdColumn=\", clusterIdColumn, \"labelColumn=\", labelColumn,\n",
    "                      \"maxWords=\", maxWords, \"useClusterMaker=\", useClusterMaker)\n",
    "    commandsGET(annotate)\n",
    "    \n",
    "    annotate <- paste(\"autoannotate layout\", \"network=\", network,\n",
    "                      \"layout=\", \"grid\")\n",
    "    commandsGET(annotate)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# General Network and Cluster analysis and evaluation informtion\n",
    "\n",
    "base_net = list()\n",
    "clustered_net = list()\n",
    "\n",
    "analyze_cmd <- paste(\"analyzer analyze directed= false\")\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    \n",
    "    network = paste(\"Proteogenomic_net_\", idx, sep = \"\")\n",
    "    setCurrentNetwork(network = network)\n",
    "    \n",
    "    base_net[[idx]] <- commandsGET(analyze_cmd)\n",
    "    \n",
    "    network = paste(\"Proteogenomic_net_\", idx, \"--clustered\", sep = \"\")\n",
    "    setCurrentNetwork(network = network)\n",
    "    \n",
    "    clustered_net[[idx]] <- commandsGET(analyze_cmd)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving the results\n",
    "\n",
    "ending = paste(species, ont_domain, \"-annotated_by\", labelColumn, sep = \"_\")\n",
    "AddrnName = paste(\"../../Results/Cy_networks/\",\n",
    "                      \"Network_session_\", ending, sep = \"\")\n",
    "saveSession( filename = AddrnName)\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    network = paste(\"Proteogenomic_net_\", idx, \"--clustered\", sep = \"\")\n",
    "    AddrnName = paste(\"../../Results/Cy_network_images/\",\n",
    "                      \"Network_images_\", ending, idx, sep = \"\")\n",
    "    exportImage(\n",
    "        filename = AddrnName,\n",
    "        type = image_type,       # PNG (default), JPEG, PDF, SVG, PS (PostScript).\n",
    "        resolution = 600,   # (numeric) The resolution of the exported image, in DPI. \n",
    "                            # Valid only for bitmap formats. The possible values are:\n",
    "                            # 72 (default), 100, 150, 300, 600.\n",
    "        network = network\n",
    "        )\n",
    "    \n",
    "    AddrnName = paste(\"../../Results/Cy_networks/\",\n",
    "                      \"Network_\", ending, idx, sep = \"\")\n",
    "    exportNetwork(\n",
    "        filename = AddrnName, # Full path or path relavtive to current working directory\n",
    "                              # in addition to the name of the file. \n",
    "        type = network_type,         # File type. SIF (default), CX, cyjs, graphML, NNF, xGMML.\n",
    "        network = network     # (optional) Name or SUID of a network or view. Default is the \"current\" network active in Cytoscape.\n",
    "    )\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "sessionInfo()\n",
    "\n",
    "R version 3.6.3 (2020-02-29)\n",
    "Platform: x86_64-conda_cos6-linux-gnu (64-bit)\n",
    "\n",
    "Matrix products: default\n",
    "BLAS/LAPACK: ../anaconda3/envs/Bioconda/lib/libopenblasp-r0.3.10.so\n",
    "\n",
    "locale:\n",
    " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
    " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
    " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
    " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
    " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
    "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
    "\n",
    "attached base packages:\n",
    "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
    "\n",
    "other attached packages:\n",
    "[1] RColorBrewer_1.1-2 RCy3_2.6.0        \n",
    "\n",
    "loaded via a namespace (and not attached):\n",
    " [1] igraph_1.2.4.1      graph_1.64.0        Rcpp_1.0.1         \n",
    " [4] magrittr_1.5        BiocGenerics_0.32.0 uuid_0.1-2         \n",
    " [7] R6_2.4.0            httr_1.4.0          tools_3.6.3        \n",
    "[10] parallel_3.6.3      R.oo_1.22.0         htmltools_0.3.6    \n",
    "[13] digest_0.6.18       crayon_1.3.4        RJSONIO_1.3-1.1    \n",
    "[16] IRdisplay_0.7.0     repr_0.19.2         base64enc_0.1-3    \n",
    "[19] R.utils_2.8.0       curl_3.3            IRkernel_0.8.15    \n",
    "[22] evaluate_0.13       pbdZMQ_0.3-3        compiler_3.6.3     \n",
    "[25] R.methodsS3_1.7.1   stats4_3.6.3        XML_3.98-1.19      \n",
    "[28] jsonlite_1.6        pkgconfig_2.0.2\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
