{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Editor : AJ\n",
    "\n",
    "######### STRING API - R implementation #########\n",
    "\n",
    "# Document and resource used : \n",
    "#\n",
    "#        -STRING API ref.\n",
    "#        -https://string-db.org/cgi/help.pl?subpage=api%23getting-the-string-network-interactions\n",
    "#\n",
    "\n",
    "############\n",
    "# purpose : analysis of protein-protein interactions of genes by utilizing STRING database\n",
    "#           and adding String proteomics information to previous genomic analysis information\n",
    "# parameters : csv files (containing EnsemblGeneID of the differentially expressed genes ),\n",
    "#              number of conditions, species, ontology domain, STRING Api's input data attributes\n",
    "# return value : \n",
    "#                  - Tabels of proteogenomic informations created by adding STRING proteomics information\n",
    "#                  to previous genomic analysis information\n",
    "#                  - STRING interaction network for identified proteins (Containing stringId_A (for protein A),\n",
    "#                  stringId_B for (protein B), common protein name (protein A), common protein name (protein B),\n",
    "#                  NCBI taxon identifier, combined score, gene neighborhood score, gene fusion score\n",
    "#                  phylogenetic profile score, coexpression score, experimental score, database score\n",
    "#                  and , textmining score)\n",
    "# details :\n",
    "#           - Species, ontology domain, addrress of the directory which contains genomic information \n",
    "#           should be provided in the section : \n",
    "#           ##  Setting initil Analysis' info  ##\n",
    "#           - Data should be provided as a .csv file containing ensembl gene IDs\n",
    "#\n",
    "############"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading needed Libraries\n",
    "\n",
    "libs = c(\"httr\", \"stringi\", \"stringr\", \"caret\", \"data.table\")\n",
    "\n",
    "for (i in libs){\n",
    "  if( !is.element(i, .packages(all.available = TRUE)) ) {\n",
    "    install.packages(i)\n",
    "  }\n",
    "  library(i,character.only = TRUE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating related directories\n",
    "\n",
    "dir.create(file.path(\"../../Results/\",\"Proteomics_interaction_tables\"), recursive = TRUE, showWarnings = FALSE)\n",
    "dir.create(file.path(\"../../Results/\",\"Proteogenomic_info\"), recursive = TRUE, showWarnings = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creatnig a 'named list' containing mapping between NCBI species Ids and\n",
    "# supported species to be used for downloading proteomics data from STRING \n",
    "\n",
    "NCBI_id_list <- list(9606,10088,10114,4932,7227,3701,7955,10224,27592,9031,9615,9823,\n",
    "                     9544,83333,8353,7164,9598,36329,83334,32)\n",
    "\n",
    "list_names <- c(\"human\",\"mouse\",\"rat\",\"yeast\",\"fly\",\"arabidopsis\",\"zebrafish\",\"worm\",\"bovinae\",\n",
    "                \"chicken\",\"canine\",\"pig\",\"rhesus\",\"ecoli_k12\",\"xenopus\",\"anopheles\",\"chimp\",\n",
    "                \"malaria\",\"ecoli_sakai\",\"myxococcus\")\n",
    "\n",
    "names(NCBI_id_list) <- list_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##       Setting initil info       ##\n",
    "#####################################\n",
    "\n",
    "# define the prfered species' name from the names in list_names \n",
    "species = \"fly\"\n",
    "\n",
    "# define the prfered ontology domain\n",
    "ont_domain = \"biological_process\"   #  possible values : \"cellular_component\",\"molecular_function\",\n",
    "                                                                                                                #\" biological_process\"\n",
    "\n",
    "##   Setting the input data info   ##\n",
    "#####################################\n",
    "\n",
    "# Defining number of conditions (treatments) except the base condition,\n",
    "# and intial STRING parameters\n",
    "treat_count = 3\n",
    "\n",
    "##  Setting STRING input data info ##\n",
    "#####################################\n",
    "\n",
    "string_api_url = \"https://string-db.org/api\"\n",
    "#output_format = \"tsv-no-header\"\n",
    "output_format = \"tsv\"\n",
    "id_method = \"get_string_ids\"\n",
    "net_method = \"network\"\n",
    "\n",
    "# required_score and additional nodes are not used if default is \"true\"\n",
    "default = \"true\"      # values : \"true\", \"false\"\n",
    "required_score = 1000 # threshold of significance to include a interaction,\n",
    "                      # a number between 0 and 1000 (default depends on the network)\n",
    "add_nodes = 0         # adds a number of proteins with to the network \n",
    "                      # based on their confidence score\n",
    "\n",
    "NCBI_species_Id=NCBI_id_list[[species]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data from file\n",
    "\n",
    "gene_dfl = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    gene_dfl[[idx]] <- read.table(file = paste(\"../../Results/Enrichment_tabels/\", species,\n",
    "                                               \"/final_df_\", ont_domain,\"_\", idx, \".csv\", sep = \"\")\n",
    "                                  ,sep = \",\", header = TRUE,  stringsAsFactors = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set parameters for STRING api identifier Mapping \n",
    "\n",
    "id_params = list()\n",
    "lid_params = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    id_params$identifiers = paste(gene_dfl[[idx]]$EnsemblGeneID, collapse = \"\\r\")\n",
    "    id_params$species = NCBI_species_Id  # species NCBI identifier \n",
    "    id_params$limit = 1 # only one (best) identifier per input protein\n",
    "    id_params$echo_query = 1 # see your input identifiers in the output\n",
    "    #id_params$caller_identity\" : \"app_name\" # your app name\n",
    "    \n",
    "    lid_params[[idx]] <- id_params\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Construct URL for STRING identifier mapping api request\n",
    "\n",
    "id_request_url = paste(string_api_url,output_format,id_method, sep = \"/\")\n",
    "\n",
    "## Call STRING\n",
    "\n",
    "lid_results = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    id_results <- POST(id_request_url, body = lid_params[[idx]])\n",
    "    lid_results[[idx]] <- id_results\n",
    "    content(lid_results[[idx]], as = \"text\", encoding = \"UTF-8\") # automatically parses JSON\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Define function to read and parse the tsv results from STRING api\n",
    "\n",
    "string_res <- function(results){\n",
    "    \n",
    "    string_result = list()\n",
    "    check = 1\n",
    "    \n",
    "    lines = str_split(results, \"\\n\")\n",
    "    len = lengths(lines)-1\n",
    "    \n",
    "    lines = unlist(lines, use.names = FALSE)\n",
    "    \n",
    "    for (idx in 1:len){\n",
    "        #string_result[[idx]] <- unlist(str_split(lines[[idx]], \"\\t\"), use.names = FALSE)\n",
    "        if (check == 1) {\n",
    "            col_name = unlist(str_split(lines[[idx]], \"\\t\"), use.names = FALSE)\n",
    "        } else {\n",
    "            data_row = unlist(str_split(lines[[idx]], \"\\t\"), use.names = FALSE)\n",
    "            string_result[[idx]] <- data_row\n",
    "        }\n",
    "        check = 0\n",
    "    }\n",
    "    df = as.data.frame(t(as.data.table(string_result)))\n",
    "    colnames(df) <- col_name\n",
    "    return(df)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parsing the results from identifier mapping api and creating a dataframe\n",
    "\n",
    "lid_parsed_results = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    lid_parsed_results[[idx]] = string_res(content(lid_results[[idx]], as = \"text\", encoding = \"UTF-8\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating a new data frame by adding String proteomics information \n",
    "# to previous genomic analysis information\n",
    "\n",
    "full_df = list()\n",
    "for (idx in 1:treat_count){\n",
    "    full_df[[idx]] = merge(gene_dfl[[idx]], lid_parsed_results[[idx]], by.x = \"ID\", by.y = \"queryItem\")\n",
    "    full_df[[idx]]$queryIndex <- NULL\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set parameters for STRING's interaction network's api\n",
    "\n",
    "net_params = list()\n",
    "lnet_params = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    \n",
    "    if (default == \"false\"){\n",
    "        net_params$required_score = required_score\n",
    "        net_params$add_nodes = net_params        \n",
    "    }\n",
    "    \n",
    "    net_params$identifiers = paste(lid_parsed_results[[idx]]$stringId, collapse = \"%0d\")\n",
    "    net_params$species = NCBI_species_Id  # species NCBI identifier\n",
    "    #net_params$caller_identity\" : \"app_name\" # your app name\n",
    "    \n",
    "    lnet_params[[idx]] <- net_params\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Construct URL for STRING interaction network's api request\n",
    "\n",
    "net_request_url = paste(string_api_url,output_format,net_method, sep = \"/\")\n",
    "\n",
    "## Call STRING\n",
    "\n",
    "lnet_results = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    net_results <- POST(net_request_url, body = lnet_params[[idx]])\n",
    "    lnet_results[[idx]] <- net_results\n",
    "    content(lnet_results[[idx]], as = \"text\", encoding = \"UTF-8\") # automatically parses JSON\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parsing the results from interaction network's api and creating a dataframe\n",
    "col_name = c(\"stringId_A\", \"stringId_B\", \"preferredName_A\", \"preferredName_B\", \"ncbiTaxonId\",\n",
    "            \"combined_score\", \"neighborhood_score\", \"fusion_score\", \"phylogenetic_profile_score\",\n",
    "            \"coexpression_score\", \"experimental_score\", \"database_score\", \"textmining_score\")\n",
    "\n",
    "lnet_parsed_results = list()\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    lnet_parsed_results[[idx]] = string_res(content(lnet_results[[idx]], as = \"text\", encoding = \"UTF-8\"))\n",
    "    colnames(lnet_parsed_results[[idx]]) <- col_name\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving the STRING api query results\n",
    "\n",
    "for (idx in 1:treat_count){\n",
    "    ending = paste(species, ont_domain, sep = \"_\")\n",
    "    write.csv(full_df[[idx]], paste(\"../../Results/Proteogenomic_info/\", ending,\n",
    "                                           \"_proteogenomic_info_\", idx, \".csv\", sep = \"\"), row.names = FALSE)\n",
    "    \n",
    "    write.csv(lnet_parsed_results[[idx]], paste(\"../../Results/Proteomics_interaction_tables/\",\n",
    "                                    ending, \"_Proteomics_interaction_table_\", idx, \".csv\", sep = \"\"), row.names = FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### END"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "sessionInfo()\n",
    "\n",
    "R version 3.6.3\n",
    "Platform: x86_64-conda_cos6-linux-gnu (64-bit)\n",
    "\n",
    "Matrix products: default\n",
    "BLAS/LAPACK: ../anaconda3/envs/Bioconda/lib/libopenblasp-r0.3.10.so\n",
    "\n",
    "locale:\n",
    " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
    " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
    " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
    " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
    " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
    "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
    "\n",
    "attached base packages:\n",
    "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
    "\n",
    "other attached packages:\n",
    "[1] data.table_1.12.2 caret_6.0-83      ggplot2_3.1.1     lattice_0.20-38  \n",
    "[5] stringr_1.4.0     stringi_1.4.6     httr_1.4.0       \n",
    "\n",
    "loaded via a namespace (and not attached):\n",
    " [1] pbdZMQ_0.3-3       tidyselect_0.2.5   repr_0.19.2        purrr_0.3.2       \n",
    " [5] reshape2_1.4.3     splines_3.6.3      colorspace_1.4-1   generics_0.0.2    \n",
    " [9] stats4_3.6.3       htmltools_0.3.6    base64enc_0.1-3    survival_2.44-1.1 \n",
    "[13] prodlim_2018.04.18 rlang_0.3.4        ModelMetrics_1.2.2 pillar_1.3.1      \n",
    "[17] glue_1.3.1         withr_2.1.2        uuid_0.1-2         foreach_1.4.4     \n",
    "[21] plyr_1.8.4         lava_1.6.5         timeDate_3043.102  munsell_0.5.0     \n",
    "[25] gtable_0.3.0       recipes_0.1.5      codetools_0.2-16   evaluate_0.13     \n",
    "[29] curl_3.3           class_7.3-15       IRdisplay_0.7.0    Rcpp_1.0.1        \n",
    "[33] scales_1.0.0       IRkernel_0.8.15    ipred_0.9-8        jsonlite_1.6      \n",
    "[37] digest_0.6.18      dplyr_0.8.0.1      grid_3.6.3         tools_3.6.3       \n",
    "[41] magrittr_1.5       lazyeval_0.2.2     tibble_2.1.1       crayon_1.3.4      \n",
    "[45] pkgconfig_2.0.2    MASS_7.3-51.3      Matrix_1.2-17      lubridate_1.7.4   \n",
    "[49] gower_0.2.0        assertthat_0.2.1   iterators_1.0.10   R6_2.4.0          \n",
    "[53] rpart_4.1-15       nnet_7.3-12        nlme_3.1-139       compiler_3.6.3    \n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
